gilbreth_grasp_planning/scripts/robot_execution_goal_abortion.py

```diff
@@ -0,0 +1,283 @@
+                pick_traj = self.motion_plan(self.arm_group,target_tool_pose.pick_pose.pose,ARM_CONTROLLER,RAIL_CONTROLLER)
+            
+                if pick_traj:
+                    pick_dur = pick_traj.joint_trajectory.points[-1].time_from_start
+                    pick_deadline = copy.deepcopy(target_tool_pose.pick_pose.header.stamp)
+                    if (rospy.Time.now() + pick_dur >pick_deadline):
+                        rospy.logerr("Can not reach pick pose on time. Drop this picking assignment.")
+                    else:
+                        ## Wait to execute robot to pick item
+                        rospy.loginfo("------Waiting to pick object-----")
+                        current_time = rospy.Time.now()
+                        wait_dur = pick_deadline.to_sec()-current_time.to_sec()-pick_dur.to_sec()
+                        rospy.sleep(wait_dur)
+
+                        rospy.loginfo("Enable Gripper.")
+                        if self.control_gripper(True) and self.arm_group.execute(pick_traj):
+                            rospy.loginfo("======Check object attachment======")
+                            if self.gripper_attached(pick_deadline): 
+                                #rospy.sleep(0.5)                  
+                                rospy.loginfo("======Go to retreat pose======")
+                                retreat_traj = self.motion_plan(self.arm_group,target_tool_pose.pick_retreat.pose,ARM_CONTROLLER,RAIL_CONTROLLER)
+
+                                if self.move_robot(self.arm_group,retreat_traj):
+                                    rospy.loginfo("======Go to place pose======")   
+                                    place_traj = self.motion_plan(self.rail_group,target_tool_pose.place_pose.pose,RAIL_CONTROLLER,ARM_CONTROLLER)
+
+                                    if self.move_robot(self.rail_group,place_traj):
+                                        rospy.loginfo("Disable Gripper")
+                                        if self.control_gripper(False):
+                                            rospy.loginfo("======Successsfully placed object======")
+ 
+            self.tool_poses_list.pop(0)
+            self.control_gripper(False)
+            self.EXECUTE = False        
+
+def main(args):
+    moveit_commander.roscpp_initialize(sys.argv)
+    rospy.init_node('robot_execution',anonymous=True)
+    rate = rospy.Rate(10)
+
+    rospy.loginfo("Waiting for move group")
+    if waitForMoveGroup():
+      rospy.loginfo("Found move group node, proceeding")
+    else:
+      rospy.logerr("Timed out waiting for move group node, exiting ...")
+      sys.exit(-1)
+
+    rb_exec = RobotExecution()
+    rb_exec.goto_waiting_pose()
+
+    while not rospy.is_shutdown():
+        rb_exec.execute_robot()
+        rate.sleep()
+
+
+if __name__=='__main__':
+    try:
+        main(sys.argv)
+    except rospy.ROSInterruptException:
+        pass
+
```

Comment:
How is this script being used? The launch file above comments out the line where this gets called.

Reply:
If you use this script to execute the robot. You can observe a weird behavior: 
After the robot successfully picks up the object, it starts to move to pick retreat pose. This execution process fails very often due to a GOAL_TOLERANCE_VIOLATION warning. This violation leads to an abortion of execution. 
I changed the controller yaml file to resolve the problem. Increasing the goal tolerance parameter will somehow decreases the chance of goal tolerance violation. But the abortion still happens very often.

Reply:
I don't know if it has something to do with the ros controller #317 issue. https://github.com/ros-controls/ros_controllers/issues/317